cibuildwheel/logger.py

```diff
@@ -0,0 +1,259 @@
+    'win32': 'Windows 32bit',
+    'win_amd64': 'Windows 64bit',
+    'macosx_x86_64': 'macOS x86_64',
+}
+
+
+class Logger:
+    fold_mode: str
+    colors_enabled: bool
+    unicode_enabled: bool
+    active_build_identifier: Optional[str] = None
+    build_start_time: Optional[float] = None
+    step_start_time: Optional[float] = None
+    active_fold_group_name: Optional[str] = None
+
+    def __init__(self):
+        if sys.platform == 'win32' and hasattr(sys.stdout, 'reconfigure'):
+            # the encoding on Windows can be a 1-byte charmap, but all CIs
+            # support utf8, so we hardcode that
+            sys.stdout.reconfigure(encoding='utf8')
+
+        self.unicode_enabled = file_supports_unicode(sys.stdout)
+
+        if 'AZURE_HTTP_USER_AGENT' in os.environ:
+            self.fold_mode = 'azure'
+            self.colors_enabled = True
+
+        elif 'GITHUB_ACTIONS' in os.environ:
+            self.fold_mode = 'github'
+            self.colors_enabled = True
+
+        elif 'TRAVIS' in os.environ:
+            self.fold_mode = 'travis'
+            self.colors_enabled = True
+
+        elif 'APPVEYOR' in os.environ:
+            self.fold_mode = 'disabled'
+            self.colors_enabled = True
+
+        else:
+            self.fold_mode = 'disabled'
+            self.colors_enabled = file_supports_color(sys.stdout)
+
+    def build_start(self, identifier: str):
+        self.step_end()
+        c = self.colors
+        description = build_description_from_identifier(identifier)
+        print()
+        print(f'{c.bold}{c.blue}Building {identifier} wheel{c.end}')
+        print(f'{description}')
+        print()
+
+        self.build_start_time = time.time()
+        self.active_build_identifier = identifier
+
+    def build_end(self):
+        assert self.build_start_time is not None
+        assert self.active_build_identifier is not None
+        self.step_end()
+
+        c = self.colors
+        s = self.symbols
+        duration = time.time() - self.build_start_time
+
+        print()
+        print(f'{c.green}{s.done} {c.end}{self.active_build_identifier} finished in {duration:.2f}s')
+        self.build_start_time = None
+        self.active_build_identifier = None
+
+    def step(self, step_description: str):
+        self.step_end()
+        self.step_start_time = time.time()
+        self._start_fold_group(step_description)
+
+    def step_end(self, success=True):
+        if self.step_start_time is not None:
+            self._end_fold_group()
+            c = self.colors
+            s = self.symbols
+            duration = time.time() - self.step_start_time
+            if success:
+                print(f'{c.green}{s.done} {c.end}{duration:.2f}s'.rjust(78))
+            else:
+                print(f'{c.red}{s.error} {c.end}{duration:.2f}s'.rjust(78))
+
+            self.step_start_time = None
+
+    def error(self, error: Union[Exception, str]):
+        self.step_end(success=False)
+        print()
```

Comment:
Same here, about the stdout vs. stderr, and redirection.

Reply:
Are you suggesting stderr might be more appropriate for an error message? I can see the argument, but pragmatically speaking I don't think users will see any difference, and because of the potential ordering issues I'd rather just stick to stdout since it's all going to a CI log anyway. Does that sound okay to you?

Reply:
Yes, but yes :-)